File: time-client/android/app/src/main/java/com/time/grpc/TimeGrpcModule.java
package com.time.grpc;

import android.util.Log;
import com.facebook.react.bridge.*;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.google.protobuf.Empty;
import com.google.protobuf.Timestamp;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

/**
 * React Native bridge module for Time gRPC client
 * Provides JavaScript interface to native gRPC functionality
 */
public class TimeGrpcModule extends ReactContextBaseJavaModule {
    private static final String TAG = "TimeGrpcModule";
    private static final String MODULE_NAME = "TimeGrpc";
    
    private TimeGrpcClient grpcClient;
    private ExecutorService executorService;
    private boolean isStreaming = false;
    
    public TimeGrpcModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.executorService = Executors.newCachedThreadPool();
    }
    
    @Override
    public String getName() {
        return MODULE_NAME;
    }
    
    /**
     * Initialize gRPC client with server configuration
     */
    @ReactMethod
    public void initialize(String host, int port, boolean useTls, Promise promise) {
        try {
            Log.d(TAG, "Initializing gRPC client: " + host + ":" + port + " (TLS: " + useTls + ")");
            
            if (grpcClient != null) {
                grpcClient.shutdown();
            }
            
            grpcClient = new TimeGrpcClient(host, port, useTls);
            
            // Test connection
            TimeResponse response = grpcClient.getCurrentTime();
            
            WritableMap result = Arguments.createMap();
            result.putString("status", "connected");
            result.putString("serverTime", response.getFormattedTime());
            result.putDouble("unixTimestamp", response.getUnixTimestamp());
            result.putString("timezone", response.getTimezone());
            
            promise.resolve(result);
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize gRPC client", e);
            promise.reject("INIT_ERROR", "Failed to initialize gRPC client: " + e.getMessage(), e);
        }
    }
    
    /**
     * Get current server time
     */
    @ReactMethod
    public void getCurrentTime(Promise promise) {
        if (grpcClient == null) {
            promise.reject("NOT_INITIALIZED", "gRPC client not initialized");
            return;
        }
        
        executorService.execute(() -> {
            try {
                TimeResponse response = grpcClient.getCurrentTime();
                
                WritableMap result = Arguments.createMap();
                result.putString("formattedTime", response.getFormattedTime());
                result.putDouble("unixTimestamp", response.getUnixTimestamp());
                result.putString("timezone", response.getTimezone());
                result.putBoolean("isUtc", response.getIsUtc());
                
                promise.resolve(result);
            } catch (Exception e) {
                Log.e(TAG, "Failed to get current time", e);
                promise.reject("GRPC_ERROR", "Failed to get current time: " + e.getMessage(), e);
            }
        });
    }
    
    /**
     * Get time with specific timezone
     */
    @ReactMethod
    public void getTimeWithTimezone(String timezone, Promise promise) {
        if (grpcClient == null) {
            promise.reject("NOT_INITIALIZED", "gRPC client not initialized");
            return;
        }
        
        executorService.execute(() -> {
            try {
                TimeResponse response = grpcClient.getTimeWithTimezone(timezone);
                
                WritableMap result = Arguments.createMap();
                result.putString("formattedTime", response.getFormattedTime());
                result.putDouble("unixTimestamp", response.getUnixTimestamp());
                result.putString("timezone", response.getTimezone());
                result.putBoolean("isUtc", response.getIsUtc());
                
                promise.resolve(result);
            } catch (Exception e) {
                Log.e(TAG, "Failed to get time with timezone", e);
                promise.reject("GRPC_ERROR", "Failed to get time with timezone: " + e.getMessage(), e);
            }
        });
    }
    
    /**
     * Get time statistics
     */
    @ReactMethod
    public void getTimeStats(Promise promise) {
        if (grpcClient == null) {
            promise.reject("NOT_INITIALIZED", "gRPC client not initialized");
            return;
        }
        
        executorService.execute(() -> {
            try {
                TimeStatsResponse response = grpcClient.getTimeStats();
                
                WritableMap result = Arguments.createMap();
                result.putString("serverVersion", response.getServerVersion());
                result.putDouble("uptimeSeconds", response.getUptimeSeconds());
                result.putDouble("totalRequests", response.getTotalRequests());
                result.putDouble("averageResponseTimeMs", response.getAverageResponseTimeMs());
                
                promise.resolve(result);
            } catch (Exception e) {
                Log.e(TAG, "Failed to get time stats", e);
                promise.reject("GRPC_ERROR", "Failed to get time stats: " + e.getMessage(), e);
            }
        });
    }
    
    /**
     * Start streaming time updates
     */
    @ReactMethod
    public void startTimeStream() {
        if (grpcClient == null) {
            Log.e(TAG, "gRPC client not initialized");
            return;
        }
        
        if (isStreaming) {
            Log.w(TAG, "Time stream already active");
            return;
        }
        
        isStreaming = true;
        Log.d(TAG, "Starting time stream");
        
        executorService.execute(() -> {
            grpcClient.streamTime(new TimeGrpcClient.TimeStreamCallback() {
                @Override
                public void onTimeUpdate(TimeUpdate update) {
                    WritableMap eventData = Arguments.createMap();
                    eventData.putString("formattedTime", update.getFormattedTime());
                    eventData.putDouble("unixTimestamp", update.getUnixTimestamp());
                    eventData.putString("timezone", update.getTimezone());
                    eventData.putInt("updateIntervalMs", update.getUpdateIntervalMs());
                    
                    sendEvent("TimeUpdate", eventData);
                }
                
                @Override
                public void onError(Throwable error) {
                    Log.e(TAG, "Time stream error", error);
                    isStreaming = false;
                    
                    WritableMap eventData = Arguments.createMap();
                    eventData.putString("error", error.getMessage());
                    
                    sendEvent("TimeStreamError", eventData);
                }
                
                @Override
                public void onCompleted() {
                    Log.d(TAG, "Time stream completed");
                    isStreaming = false;
                    sendEvent("TimeStreamCompleted", null);
                }
            });
        });
    }
    
    /**
     * Stop streaming time updates
     */
    @ReactMethod
    public void stopTimeStream() {
        if (!isStreaming) {
            Log.w(TAG, "Time stream not active");
            return;
        }
        
        Log.d(TAG, "Stopping time stream");
        isStreaming = false;
        
        // Note: gRPC streams don't have a direct way to cancel from client side
        // The server should handle this gracefully
        sendEvent("TimeStreamStopped", null);
    }
    
    /**
     * Check connection status
     */
    @ReactMethod
    public void getConnectionStatus(Promise promise) {
        if (grpcClient == null) {
            WritableMap result = Arguments.createMap();
            result.putString("status", "NOT_INITIALIZED");
            result.putBoolean("connected", false);
            promise.resolve(result);
            return;
        }
        
        WritableMap result = Arguments.createMap();
        result.putString("status", grpcClient.getConnectionState());
        result.putBoolean("connected", grpcClient.isConnected());
        result.putBoolean("streaming", isStreaming);
        
        promise.resolve(result);
    }
    
    /**
     * Shutdown gRPC client
     */
    @ReactMethod
    public void shutdown() {
        Log.d(TAG, "Shutting down gRPC module");
        
        isStreaming = false;
        
        if (grpcClient != null) {
            grpcClient.shutdown();
            grpcClient = null;
        }
        
        if (executorService != null) {
            executorService.shutdown();
            try {
                if (!executorService.awaitTermination(5, TimeUnit.SECONDS)) {
                    executorService.shutdownNow();
                }
            } catch (InterruptedException e) {
                executorService.shutdownNow();
                Thread.currentThread().interrupt();
            }
        }
    }
    
    /**
     * Send event to JavaScript
     */
    private void sendEvent(String eventName, WritableMap params) {
        ReactApplicationContext reactContext = getReactApplicationContext();
        if (reactContext.hasActiveCatalystInstance()) {
            reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
        }
    }
    
    @Override
    public void onCatalystInstanceDestroy() {
        super.onCatalystInstanceDestroy();
        shutdown();
    }
}

File: time-client/ios/Time/TimeGrpcModule.swift
import Foundation
import React
import GRPC

/**
 * React Native bridge module for Time gRPC client
 * Provides JavaScript interface to native gRPC functionality
 */
@objc(TimeGrpcModule)
class TimeGrpcModule: RCTEventEmitter {
    
    private var grpcClient: TimeGrpcClient?
    private var isStreaming = false
    private let queue = DispatchQueue(label: "com.time.grpc", qos: .userInitiated)
    
    // MARK: - RCTEventEmitter Overrides
    
    override func supportedEvents() -> [String]! {
        return ["TimeUpdate", "TimeStreamError", "TimeStreamCompleted", "TimeStreamStopped"]
    }
    
    override static func requiresMainQueueSetup() -> Bool {
        return false
    }
    
    // MARK: - React Native Methods
    
    /**
     * Initialize gRPC client with server configuration
     */
    @objc
    func initialize(_ host: String, port: NSNumber, useTls: Bool, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
        queue.async { [weak self] in
            do {
                print("TimeGrpcModule: Initializing gRPC client: \(host):\(port) (TLS: \(useTls))")
                
                // Clean up existing client
                self?.grpcClient?.shutdown()
                
                // Create new client
                let client = try TimeGrpcClient(host: host, port: port.intValue, useTLS: useTls)
                self?.grpcClient = client
                
                // Test connection
                let response = try client.getCurrentTime()
                
                let result: [String: Any] = [
                    "status": "connected",
                    "serverTime": response.formattedTime,
                    "unixTimestamp": response.unixTimestamp,
                    "timezone": response.timezone
                ]
                
                DispatchQueue.main.async {
                    resolver(result)
                }
                
            } catch {
                print("TimeGrpcModule: Failed to initialize gRPC client: \(error)")
                DispatchQueue.main.async {
                    rejecter("INIT_ERROR", "Failed to initialize gRPC client: \(error.localizedDescription)", error)
                }
            }
        }
    }
    
    /**
     * Get current server time
     */
    @objc
    func getCurrentTime(_ resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
        guard let client = grpcClient else {
            rejecter("NOT_INITIALIZED", "gRPC client not initialized", nil)
            return
        }
        
        queue.async { [weak self] in
            do {
                let response = try client.getCurrentTime()
                
                let result: [String: Any] = [
                    "formattedTime": response.formattedTime,
                    "unixTimestamp": response.unixTimestamp,
                    "timezone": response.timezone,
                    "isUtc": response.isUtc
                ]
                
                DispatchQueue.main.async {
                    resolver(result)
                }
                
            } catch {
                print("TimeGrpcModule: Failed to get current time: \(error)")
                DispatchQueue.main.async {
                    rejecter("GRPC_ERROR", "Failed to get current time: \(error.localizedDescription)", error)
                }
            }
        }
    }
    
    /**
     * Get time with specific timezone
     */
    @objc
    func getTimeWithTimezone(_ timezone: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
        guard let client = grpcClient else {
            rejecter("NOT_INITIALIZED", "gRPC client not initialized", nil)
            return
        }
        
        queue.async { [weak self] in
            do {
                let response = try client.getTimeWithTimezone(timezone)
                
                let result: [String: Any] = [
                    "formattedTime": response.formattedTime,
                    "unixTimestamp": response.unixTimestamp,
                    "timezone": response.timezone,
                    "isUtc": response.isUtc
                ]
                
                DispatchQueue.main.async {
                    resolver(result)
                }
                
            } catch {
                print("TimeGrpcModule: Failed to get time with timezone: \(error)")
                DispatchQueue.main.async {
                    rejecter("GRPC_ERROR", "Failed to get time with timezone: \(error.localizedDescription)", error)
                }
            }
        }
    }
    
    /**
     * Get time statistics
     */
    @objc
    func getTimeStats(_ resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
        guard let client = grpcClient else {
            rejecter("NOT_INITIALIZED", "gRPC client not initialized", nil)
            return
        }
        
        queue.async { [weak self] in
            do {
                let response = try client.getTimeStats()
                
                let result: [String: Any] = [
                    "serverVersion": response.serverVersion,
                    "uptimeSeconds": response.uptimeSeconds,
                    "totalRequests": response.totalRequests,
                    "averageResponseTimeMs": response.averageResponseTimeMs
                ]
                
                DispatchQueue.main.async {
                    resolver(result)
                }
                
            } catch {
                print("TimeGrpcModule: Failed to get time stats: \(error)")
                DispatchQueue.main.async {
                    rejecter("GRPC_ERROR", "Failed to get time stats: \(error.localizedDescription)", error)
                }
            }
        }
    }
    
    /**
     * Start streaming time updates
     */
    @objc
    func startTimeStream() {
        guard let client = grpcClient else {
            print("TimeGrpcModule: gRPC client not initialized")
            return
        }
        
        if isStreaming {
            print("TimeGrpcModule: Time stream already active")
            return
        }
        
        isStreaming = true
        print("TimeGrpcModule: Starting time stream")
        
        queue.async { [weak self] in
            client.streamTime(callback: TimeStreamCallbackImpl(module: self))
        }
    }
    
    /**
     * Stop streaming time updates
     */
    @objc
    func stopTimeStream() {
        if !isStreaming {
            print("TimeGrpcModule: Time stream not active")
            return
        }
        
        print("TimeGrpcModule: Stopping time stream")
        isStreaming = false
        
        // Note: gRPC streams don't have a direct way to cancel from client side
        // The server should handle this gracefully
        sendEvent(withName: "TimeStreamStopped", body: nil)
    }
    
    /**
     * Check connection status
     */
    @objc
    func getConnectionStatus(_ resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
        guard let client = grpcClient else {
            let result: [String: Any] = [
                "status": "NOT_INITIALIZED",
                "connected": false
            ]
            resolver(result)
            return
        }
        
        let result: [String: Any] = [
            "status": client.connectionState,
            "connected": client.isConnected,
            "streaming": isStreaming
        ]
        
        resolver(result)
    }
    
    /**
     * Shutdown gRPC client
     */
    @objc
    func shutdown() {
        print("TimeGrpcModule: Shutting down gRPC module")
        
        isStreaming = false
        
        queue.async { [weak self] in
            self?.grpcClient?.shutdown()
            self?.grpcClient = nil
        }
    }
    
    // MARK: - Private Methods
    
    private func sendEvent(withName name: String, body: Any?) {
        sendEvent(withName: name, body: body)
    }
}

// MARK: - TimeStreamCallback Implementation

private class TimeStreamCallbackImpl: TimeStreamCallback {
    private weak var module: TimeGrpcModule?
    
    init(module: TimeGrpcModule?) {
        self.module = module
    }
    
    func onTimeUpdate(_ update: TimeUpdate) {
        let eventData: [String: Any] = [
            "formattedTime": update.formattedTime,
            "unixTimestamp": update.unixTimestamp,
            "timezone": update.timezone,
            "updateIntervalMs": update.updateIntervalMs
        ]
        
        DispatchQueue.main.async { [weak self] in
            self?.module?.sendEvent(withName: "TimeUpdate", body: eventData)
        }
    }
    
    func onError(_ error: Error) {
        print("TimeGrpcModule: Time stream error: \(error)")
        
        let eventData: [String: Any] = [
            "error": error.localizedDescription
        ]
        
        DispatchQueue.main.async { [weak self] in
            self?.module?.isStreaming = false
            self?.module?.sendEvent(withName: "TimeStreamError", body: eventData)
        }
    }
    
    func onCompleted() {
        print("TimeGrpcModule: Time stream completed")
        
        DispatchQueue.main.async { [weak self] in
            self?.module?.isStreaming = false
            self?.module?.sendEvent(withName: "TimeStreamCompleted", body: nil)
        }
    }
}